feat: add mimetype type to source object when possible (#4469)

davidgg opened the pull request:
File mimetype is filled when the file extension is known and type is not provided.

## Description
Feature to support some bad behaviours like #4469 and #4851.

## Requirements Checklist
- [X] Feature implemented / Bug fixed
- [X] If necessary, more likely in a feature request than a bug fix
  - [X] Change has been verified in an actual browser (Chome, Firefox, IE)
  - [X] Unit Tests updated or fixed
  - [ ] Docs/guides updated
  - [ ] Example created ([starter template on JSBin](http://jsbin.com/axedog/edit?html,output))
- [x] Reviewed by Two Core Contributors

gkatsev commented:
Looks like compared to master, we are gaining ~0.15KB of minified gzipped filesize:
on [this branch](https://travis-ci.org/videojs/video.js/builds/340192167#L3691)
```
┌────────────────────────┬───────────┬───────────┐
│ filename               │      size │   gzipped │
├────────────────────────┼───────────┼───────────┤
│ alt/video.novtt.js     │ 648.37 KB │ 144.99 KB │
├────────────────────────┼───────────┼───────────┤
│ alt/video.novtt.min.js │ 169.26 KB │  43.32 KB │
├────────────────────────┼───────────┼───────────┤
│ video.cjs.js           │ 637.07 KB │ 141.65 KB │
├────────────────────────┼───────────┼───────────┤
│ video.es.js            │ 636.84 KB │ 141.59 KB │
├────────────────────────┼───────────┼───────────┤
│ video.js               │ 701.28 KB │ 159.38 KB │
├────────────────────────┼───────────┼───────────┤
│ video.min.js           │ 188.47 KB │  49.73 KB │
└────────────────────────┴───────────┴───────────┘
```
and [on master](https://travis-ci.org/videojs/video.js/builds/340272195#L3823):
```
┌────────────────────────┬───────────┬───────────┐
│ filename               │      size │   gzipped │
├────────────────────────┼───────────┼───────────┤
│ alt/video.novtt.js     │ 647.29 KB │ 144.61 KB │
├────────────────────────┼───────────┼───────────┤
│ alt/video.novtt.min.js │ 168.89 KB │  43.18 KB │
├────────────────────────┼───────────┼───────────┤
│ video.cjs.js           │ 635.99 KB │ 141.27 KB │
├────────────────────────┼───────────┼───────────┤
│ video.es.js            │ 635.76 KB │ 141.22 KB │
├────────────────────────┼───────────┼───────────┤
│ video.js               │  700.2 KB │ 159.01 KB │
├────────────────────────┼───────────┼───────────┤
│ video.min.js           │  188.1 KB │  49.58 KB │
└────────────────────────┴───────────┴───────────┘
```

davidgg commented:
Hi, I've changed the code to remove some mimetypes. This hash should contain only popular mimetypes when extension differs from `video/${extension}` template. Saying 'popular' is easy to understand but not to mantain, so you probably will have to add some extensions. I implemented the ones I considered popular. 

I've changed also the filter-sources method to lighted it up a little. More optimization is posible but compromising code readability.

hartman commented:
Small observation of this patch: since matroska and mp4 are listed... what about webm videos ?

gkatsev commented:
Probably makes sense to add webm, though, I this wasn't meant as a definitive or anything and that we can add or remove things if needed.

davidgg commented:
@gkatsev @hartman That's not necessary.  As I said, only files whose mimetype differs from `video/${extension}` template should appear in the hash. That is not the case for webm, as its mimetype is `video/webm` (https://www.webmproject.org/docs/container/). Right now video.js is already using a fallback to know if the browser can handle a video, that's the reason we need the hash. It uses that template when src has not a type property, as you can see here: 

https://github.com/videojs/video.js/blob/master/src/js/tech/html5.js#L1772-L1777

This fallback covers most of the cases but sometimes is wrong so we are using the mimetypes hash to fill type property for known problematic extensions.

**Additional**.[ Is the Video source type tag optional? If it is should I put the src inline rather than nested](https://stackoverflow.com/questions/34506704/is-the-video-source-type-tag-optional-if-it-is-should-i-put-the-src-inline-rath)?

gkatsev commented:
Ah, that's great @davidgg, I totally didn't know this :)

I think with this PR we already cover the case of a `source` element without a type attribute. Since each source in an array gets run through `filterSource`. Though, I guess would be good to verify.

davidgg commented:
I didn't manually check it but you can pass an array of sources to filter, that case is automatically covered in the new tests. Every item in the array is processed.

gkatsev commented:
Yeah, source elements are turned into an array of source objects when the player is initialized, so, we're probably all set. At the same time, I haven't seen anyone use a source element and not provide a type yet, even if it is technically optional :)

gkatsev commented:
Also, this is good to go but there are a few patch fixes I'd like to get in before doing another minor release, so, letting it hang out a bit.

gkatsev commented:
Thanks @davidgg!
